modules/runtime/nuxeo-runtime-migration/src/main/java/org/nuxeo/runtime/migration/MigrationServiceImpl.java

```diff
@@ -529,4 +530,36 @@ protected void atomic(String id, Consumer<KeyValueStore> consumer) {
         throw new RuntimeException("Cannot lock for write migration: " + id + ", already locked: " + currentLock);
     }
 
+    @Override
+    public Migration getMigration(String id) {
+        MigrationDescriptor descriptor = getDescriptor(XP_CONFIG, id);
+        MigrationStatus status = getStatus(id);
+        if (descriptor == null || status == null) {
+            return null;
+        }
+        return Migration.from(descriptor, status);
+    }
+
+    @Override
+    public List<Migration> getMigrations() {
+        return getDescriptors(XP_CONFIG).stream().map(Descriptor::getId).map(this::getMigration).collect(toList());
+    }
+
+    /**
+     * @implNote Runs a migration if it has exactly one available step to run.
+     */
+    @Override
+    public void probeAndRun(String id) {
+        MigrationStatus status = getStatus(id);
+        if (status == null) {
+            probeAndSetState(id);
+        }
+        var steps = Migration.from(getDescriptor(XP_CONFIG, id), status).getSteps();
```

Comment:
Shouldn't you get status in case you probe the state?

Reply:
The name was changed to probeAndRun later but it was just defined as run and meant to do so even if the state is unknown (so we probe then).
I think it was renamed probeAndRun to make the call to probeAndSetState more explicit as it has this warning:
> * THIS METHOD MAY TAKE A WHILE as it needs to get information from persistent storage.

@troger might remember better than me

Reply:
@kevinleturc is pointing that if `status` is null, you call `probeAndSetState(id);` but never get back the updated status, so status is still `null`.

Reply:
Exactly

Reply:
I made a systematic call to `probeAndSetState` then `getStatus`

Reply:
No no... why probing if there is no need to probe?

```java
MigrationStatus status = getStatus(id);
if (status == null) {
    probeAndSetState(id);
    status = getStatus(id);
}
...
```

It is stated in your method's Javadoc:
> Probes the migration state if necessary and runs a migration.

Reply:
alright